### DISCUSSION TITLE
How to automatic start scanning after nc restart?

### DISCUSSION THREAD

**Discussion by DenuxPlays** [nextcloud/recognize#821] [Q&A] [answered]
Hey,
my nextcloud restarts every day (for backup purposes) and recognize does not restart scanning all photo.
![image](https://user-images.githubusercontent.com/83671398/235354374-c99f548d-b255-401c-a237-80fab4c675ec.png)
How can I change this?
Because its very annoying to go into nextcloud and click on rescan every file ever day.

**Comment by DenuxPlays** [comment]
Still not working:
![image](https://user-images.githubusercontent.com/83671398/236295093-e5252dcb-bb82-4b2b-b2c0-24eebb17db7f.png)

Recognize is not working after nextcloud restart...

**Comment by marcelklehr** [comment]
> my nextcloud restarts every day (for backup purposes)

When a background job is killed while running it will not be executed for a minimum of 12h, afaicr. This is not up to recognize.

**Reply by ShapeShifter499** [reply]
I'm struggling to find the bit of code. Where does it dictate a minimum of 12h?

**Reply by marcelklehr** [reply]
https://github.com/nextcloud/server/blob/master/lib/private/BackgroundJob/JobList.php#L220

**Comment by marcelklehr** [comment]
> click on rescan every file ever day.

This is not a good idea, as it will add *all* files to the queue again, regardless of whether they have been scanned before.

**Reply by DenuxPlays** [reply]
Yeah I know but its the only way afaik to get recognize to work again...

**Reply by marcelklehr** [reply]
The problem IMO is killing the background job. Either you somehow avoid that or you wait 12h until nextcloud executes it again.

Why do you need to kill your nextcloud for the backup? Is maintenance mode not enough?

**Reply by DenuxPlays** [reply]
> The problem IMO is killing the background job. Either you somehow avoid that or you wait 12h until nextcloud executes it again.
> 
> 
> 
> Why do you need to kill your nextcloud for the backup? Is maintenance mode not enough?

The backup from nextcloud aio stops the container I can not change this.
But the last run wad 7 days ago and I uploaded a few pictures that didn't got scanned.

**Reply by marcelklehr** [reply]
cc @szaimen

**Reply by szaimen** [reply]
So the thing is that we stop all containers before creating the backup in order to make sure that the backup has a consistant state (like also the database and everything else). 

I currently don't see how enabling the maintenance mode would help to achieve this?

Also shouldnt we rather fix the cronjob behaviour to start its job sooner again and not having to wait 12h for it to start again?

**Reply by ShapeShifter499** [reply]
Would it be possible to add some kind of manual scan option that could be triggered manually like how https://github.com/nextcloud/previewgenerator does it?

**Reply by ShapeShifter499** [reply]
@DenuxPlays does this still occur for you? I noticed in the last few days the issue went away despite having automatic updates enabled in AIO

**Reply by marcelklehr** [reply]
> Also shouldnt we rather fix the cronjob behaviour to start its job sooner again and not having to wait 12h for it to start again?

I think this is a reasonable behavior. You don't want a broken background job to rerun every 5minutes and flood the logs with errors.

**Reply by szaimen** [reply]
Shouldnt there be a way to differentiate between broken vs not ended background jobs (jobs that took less than 12h but were ended from the outside)? If there was one, it should be possible to add a logic to restart not ended background jobs earlier...

**Reply by ShapeShifter499** [reply]
@marcelklehr forgive me as I haven't looked through all of the code. I'm not sure on the order of things that is breaking here. 

Do I understand that Recognize is processing files one by one? If the Nextcloud AIO instance is interrupted while a file is being processed, Recognize fails on next startup?

Am I just getting lucky that the last few restarts of my Nextcloud AIO install has happened between files being processed?

**Reply by marcelklehr** [reply]
I think it's rather likely that the recognize background jobs do end, just not successfully, but with an exception when the node process is killed by killing the container.

**Reply by marcelklehr** [reply]
In my opinion, when a) a container is killed ad hoc and thus b) the processes in it get killed, causing c) a background job to throw an error, in turn causing d) that background job to not be called for a while; if you don't want d) to happen you need to avoid a).

**Reply by marcelklehr** [reply]
Maybe you can enable maintenance-mode and wait for all cron.php processes to stop before killing the container. Maintenance mode will halt execution of new background jobs.

**Reply by szaimen** [reply]
> Maybe you can enable maintenance-mode and wait for all cron.php processes to stop before killing the container. Maintenance mode will halt execution of new background jobs.

I tried to listen for the container shutdown process and wait for cron to finish but it didnt work. See https://github.com/nextcloud/all-in-one/pull/3214. Maybe you have a good idea how to do that?

**Reply by szaimen** [reply]
Alao I guess we cannot really wait longer than 600s so it will not fix all issues as otherwise you might need to wait just too long until you can start a backup or start the containers again... (The instance will be down for that time)

**Reply by DenuxPlays** [reply]
> @DenuxPlays does this still occur for you? I noticed in the last few days the issue went away despite having automatic updates enabled in AIO

![image](https://github.com/nextcloud/recognize/assets/83671398/77a6cb9b-51b4-4b8b-9cd7-d6c145b631e7)
Yes still the same...

**Comment by szaimen** [comment, marked as answer]
The fix https://github.com/nextcloud/all-in-one/pull/3214 is now released with v7.8.0 Beta. Testing and feedback is welcome! See [nextcloud/all-in-one#how-to-switch-the-channel](https://github.com/nextcloud/all-in-one#how-to-switch-the-channel)

**Reply by DenuxPlays** [reply]
> The fix https://github.com/nextcloud/all-in-one/pull/3214 is now released with v7.8.0 Beta. Testing and feedback is welcome! See [nextcloud/all-in-one#how-to-switch-the-channel](https://github.com/nextcloud/all-in-one#how-to-switch-the-channel)

Nice
I am going to test this when I am home (in about 2,5h)

**Comment by DenuxPlays** [comment]
> The fix [nextcloud/all-in-one#3214](https://github.com/nextcloud/all-in-one/pull/3214) is now released with v7.8.0 Beta. Testing and feedback is welcome! See [nextcloud/all-in-one#how-to-switch-the-channel](https://github.com/nextcloud/all-in-one#how-to-switch-the-channel)

Update:
atm nothing seems to work:
![image](https://github.com/nextcloud/recognize/assets/83671398/cb835d7c-3d98-4b4f-b84a-a0a5ffaa3c75)

Tried rescanning all files, recrawl everything via the terminal...
Nothing seems to work.

**Reply by DenuxPlays** [reply]
Running Nextcloud AIO version 7.8.1 Beta

**Reply by szaimen** [reply]
Can you post the nextcloud container logs here after shutting down the container before starting it up again?

**Reply by DenuxPlays** [reply]
Just checked in and it seems to just work again
![image](https://github.com/nextcloud/recognize/assets/83671398/8d4e68ec-7cba-4b9e-8b46-adc1bd65e915)

I don't why as I haven't opened nextcloud or did anything since yesterday but who knows.
I also took a look at the nextcloud log but nothing important there...

The only thing I can imagine is they scheduled restart (every day) but I don't know.
But the restart was roughly 11h hours ago and I've synced my photos from the phone to my nextcloud instance and it should've triggered a background job.
I just keep an eye on it for the rest of the day and give an update if nothing happens.

**Reply by DenuxPlays** [reply]
Yes seems to be broken:
![image](https://github.com/nextcloud/recognize/assets/83671398/41f9f8cf-1410-4481-856c-985aee38a32c)

And the logs says that all cronjobs finished successfully:
![image](https://github.com/nextcloud/recognize/assets/83671398/7980cae9-cf16-4e71-ab02-4c53627f29da)

**Reply by DenuxPlays** [reply]
Info:
I found something interesting.
If you stop the nextcloud container and then start it again (manually) it invokes a new recognize scan:
![image](https://github.com/nextcloud/recognize/assets/83671398/2fb35731-74cb-4c04-9b2e-95116c21aeb0)

If nextcloud restarts automatically (because it's doing an backup) nothing happens.

**Reply by szaimen** [reply]
You might run into this timeout then: https://github.com/nextcloud/all-in-one/blob/ad4b48c82e56b2d0769fa78912da9839bf290b99/php/containers.json#L218
But I fear we cannot increase this because it would increase the downtime. Also having to wait up to 10min for your containers to stop already feels too long imho...

**Reply by DenuxPlays** [reply]
Yeah I totally agree...
5 minutes should be enough.
But can we make it configurable via env-vars?
Then I could at least test it or for me personal it is okay if my nextcloud isn't available for a few more minutes as I usally sleep during the downtime, but this will not apply to everyone especially if more than a few users use it.

**Reply by szaimen** [reply]
Instead of having yet another config option would the more elegant way be this:

> Shouldnt there be a way to differentiate between broken vs not ended background jobs (jobs that took less than 12h but were ended from the outside)? If there was one, it should be possible to add a logic to restart not ended background jobs earlier...

**Reply by DenuxPlays** [reply]
> Instead of having yet another config option would the more elegant way be this:
> 
> 
> 
> > Shouldnt there be a way to differentiate between broken vs not ended background jobs (jobs that took less than 12h but were ended from the outside)? If there was one, it should be possible to add a logic to restart not ended background jobs earlier...

Yes that sounds like a clean solution but it is realistic that this will be implemented any time soon?
I imagine this has to be implemented in nextcloud directly?
